### ISSUE TITLE
cgen generated wrong type inside the match case

### ISSUE THREAD

**Issue by endeveit** [vlang/v#6016]
**V version:** `V 0.1.28 280afb2`
**OS:** `macOS 10.15.6`

**What did you do?**
```v
module main

type CommonType = any_int | any_float | string

fn (c CommonType) str() string {
	match c {
		string {
			return c
		}
		else {
			return c.str()
		}
	}
}

fn as_string(val CommonType) string {
	return 'This is the string representation of "' + val.str() + '"'
}

fn main() {
	println(as_string(1))
	println(as_string(3.14))
	println(as_string('String'))
}
```

**What did you expect to see?**
```bash
$ v run main.v
This is the string representation of "1"
This is the string representation of "3.14"
This is the string representation of "String"
```
 
**What did you see instead?**
```bash
$ v run main.v
==================
^
/private/var/folders/4z/zkxqxv9n545dgkx16_d5zhmm0000gn/T/v/main.tmp.c:9470:10: error: indirection requires pointer operand ('string' (aka 'struct string') invalid)
                return **c;
                       ^~~
/private/var/folders/4z/zkxqxv9n545dgkx16_d5zhmm0000gn/T/v/main.tmp.c:9509:38: warning: integer literal is too large to be represented in a signed integer type, interpreting as unsigned [-Wimplicitly-unsigned-literal]
        _const_math__bits__max_u64 = ((u64)(18446744073709551615));
                                            ^
/private/var/folders/4z/zkxqxv9n545dgkx16_d5zhmm0000gn/T/v/main.tmp.c:9536:409: warning: integer literal is too large to be represented in a signed integer type, interpreting as unsigned [-Wimplicitly-unsigned-literal]
                ((u64)(1)), ((u64)(10)), ((u64)(100)), ((u64)(1000)), ((u64)(10000)), ((u64)(100000)), ((u64)(1000000)), ((u64)(10000000)), ((u64)(100000000)), ((u64)(1000000000)), ((u64)(10000000000)), ((u64)(100000000000)), ((u64)(1000000000000)), ((u64)(10000000000000)), ((u64)(100000000000000)), ((u64)(1000000000000000)), ((u64)(10000000000000000)), ((u64)(100000000000000000)), ((u64)(1000000000000000000)), ((u64)(10000000000000000000))}));
                                                                                                                                                                                                                                                                                                                                                                                                                                      ^
18 warnings and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

If you were not working with C interop, please raise an issue on GitHub:

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

**Comment by SleepyRoy** [comment]
I could do a quick fix (with only two lines) to make this workaround work.
```V
module main

type CommonType = any_int | any_float | string

fn (c CommonType) str() string {
	match c {
		string {
			return c
		}
		any_int {
			return c.str()
		}
		any_float {
			return c.str()
		}
	}
}

fn as_string(val CommonType) string {
	return 'This is the string representation of "' + val.str() + '"'
}

fn main() {
	println(as_string(1))
	println(as_string(3.14))
	println(as_string('String'))
}
```
However, more work with cgen for match_expr is needed to make your original code work.

**Comment by endeveit** [comment]
Hi @SleepyRoy, it didn't fix it, actually.
Which version of V do you use?
I'm on:
```bash
$ v version
V 0.1.28 81f8e91
```

**Comment by SleepyRoy** [comment]
@endeveit Yeah the fix is in my PR #6012 and not merged yet. The workaround example is added to tests.

**Comment by danieldaeschle** [comment]
works now with this code
```v
module main

type CommonType = f32 | int | string

fn (c CommonType) str() string {
	match c {
		string {
			return c
		}
		f32, int {
			return c.str()
		}
	}
}

fn as_string(val CommonType) string {
	return 'This is the string representation of "' + val.str() + '"'
}

fn main() {
	println(as_string(1))
	println(as_string(f32(3.14)))
	println(as_string('String'))
}
```
